### ISSUE TITLE
Blueprint flipping does not work

### ISSUE THREAD

**Issue by LevFendi** [Factorio-Access/FactorioAccess#208]
Blueprint flipping with the F and G keys flips the visual preview but the blueprint object does not change (even if you export and import again) and the flip effect does not apply to `build_from_cursor`. It is possible that this had never worked in the first place.

There appears to be no way to fix this without requesting an API change. For example we can ask for the addition of parameters to 'can_build_from_cursor' and 'build_from_cursor' for `flipped_horizontal` and `flipped_vertical`.

EDIT: Update. Turns out that we have `set_blueprint_entities` which means that we can manually write out a function to flip a blueprint by manipulating entity positions and rotations. It is just a lot of effort but it is doable.

**Comment by LevFendi** [comment]
For now I have added error messages in game to note that this does not work

**Comment by ahicks92** [comment]
2.0 takes this and adds the special hell of mirroring.  We can probably launch without supporting it but in general this got much harder, and also the result of placing flipped blueprints is no longer just rotate 180.

**Comment by LevFendi** [comment]
I will check sometime what the new API has for this

**Comment by EphDoering** [comment]
Only thing in 2.0.7 appears to be just a bool if an entity is mirrored.

**Comment by LevFendi** [comment]
if we are able to write it, maybe that is all we need?

**Comment by EphDoering** [comment]
the flag is on the placed entities as far as I can tell it's not on the blueprint item.

**Comment by ahicks92** [comment]
Well, ok so kind of.  You can almost certainly blueprint a flipped entity and the flag will have to come along with that, otherwise they can't support flipping on everything and you couldn't e.g. take a blueprint of a flipped blueprint and a non-flipped blueprint next to each other.

But that's not the issue.  The issue is that sometimes you need a 180 degree rotation and sometimes you need to mirror.  I don't think we have enough information out of the engine to easily tell which is which without having to reimplement their symmetry finding thing.

If there was a time to sidechannel with someone at Wube to get this into the API this is possibly our one single "we really need it" API request.  I may bump our outstanding request in a week or two.

Someone also needs to check what happens in the API if mirroring entities that don't need to be, e.g. an assembling machine or inserter.  This means both if placed/done through the UI/player actions, and specifically writing the mirroring attribute.  A lot depends on whether or not the engine normalizes those cases.

**Comment by EphDoering** [comment]
yeah I commented on #290 with that info. TLDR the engine normalizes. Only when it was necessary to relocate the fluidbox connections does the mirroring member of the entity return true. I think we could even close 290 (and maybe open an issue to support notification of flipping) and leave this issue to address flipping blueprints.

**Comment by ahicks92** [comment]
No, we can't yet.  I will drop a comment over there.

**Comment by ahicks92** [comment]
I don't see where I proposed widening the scope, but I might be being an idiot.
